### ISSUE TITLE
Modules executing in reverse post-order

### ISSUE THREAD

**Issue by guybedford** [microsoft/ChakraCore#6261]
Module execution of ES modules is based on post-order execution up the graph (with circular deduping).

The Edge behaviour is incorrectly implementing a _reverse post-order execution_.

The simple example of this is the following:

```html
<script>order=[]</script>
<script type="module">
import '/a.js';
import '/b.js';
console.log(order);
</script>
```

where `a.js` contains `order.push('a')` and `b.js` similarly.

where edge will execute `b` and then `a`, instead of `a` and then `b`!

The fix should be simple (one loop can just be reversed somewhere), but this is a really important bug to fix as many projects will write code like the following:

```js
import 'polyfill';
import 'stuff-that-needs-polyfill';
```

and without these guaranatees those code paths break down.

I would highly recommend marking this as an urgent priority.

**Comment by guybedford** [comment]
The scenario here might be a little more complex actually. For example, the following code executes dependencies in **indeterminate** order:

```html
<!doctype html>
<script>
function blobUrl (source) {
  return URL.createObjectURL(new Blob([source], { type: 'application/javascript' }));
}

window.order = [];
document.head.appendChild(Object.assign(document.createElement('script'), {
  type: 'module',
  innerHTML: `
    import '${blobUrl('order.push("a")')}';
    import '${blobUrl('order.push("b")')}';
    console.log(order);
  `
}));
</script>
```

when the ordering always should be `[a,b]`, it will vary between that and `[b,a]` roughly 50% of the time.

**Comment by guybedford** [comment]
On the other hand, the following always executes in the correct order:

```html
<!doctype html>
<script>
window.order = [];
document.head.appendChild(Object.assign(document.createElement('script'), {
  type: 'module',
  innerHTML: `
    import 'data:application/javascript,order.push("a")';
    import 'data:application/javascript,order.push("b")';
    console.log(order);
  `
}));
</script>
```

It seems to me like request timings or otherwise are possibly being allowed to affect execution order.

**Comment by fatcerberus** [comment]
Is there a specific loading order guaranteed by the specification?

**Comment by rhuanjl** [comment]
@fatcerberus yes there is.

@guybedford the module load order was wrong in versions 1.8 and 1.9 of CC it should have been fixed for 1.10 by this PR of mine and I believe that the most recent version of Edge should be using 1.10 or even 1.11 https://github.com/microsoft/ChakraCore/pull/5238

The issue with indeterminate order sounds to me like a problem with Edge not CC based on my knowledge of the CC code base but as an external contributor to CC i’ce not seen any of the Edge code so can’t say for sure.

**Comment by guybedford** [comment]
@rhuanjl I was testing on Edge 44.18362. If you think the out-of-order issue will have been resolved since then that is great to hear.

**Comment by rhuanjl** [comment]
I’m afraid I don’t know which version of CC each version of Edge uses; will need one g the Microsoft guys to comment.

**Comment by rhuanjl** [comment]
Taking another look at this - I'm wrong above. CC module load order remains wrong in the latest version. I fixed one error with my previous PR (execution of the list of modules backwards) but there were two errors.

The list of child modules is being stored internally in an internal data structure type called a BaseDictionary - which use Key, Value pairs and is sorted based on the keys - in this case the specifiers, so execution order is dependent on the specifier - the test case I added before had conveniently alphabetical module names so appeared to work.

I'll see if I can submit a fix shortly - it'll be slightly messy due to how wrapped around this BaseDictionary type the implementation is.
